model/map/src/main/java/org/keycloak/models/map/client/MapClientProviderFactory.java

```diff
@@ -16,72 +16,55 @@
 import org.keycloak.models.RoleModel;
-import org.keycloak.provider.ProviderEvent;
-import org.keycloak.provider.ProviderEventListener;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentMap;
+import org.keycloak.provider.InvalidationHandler;
+import org.keycloak.provider.InvalidationHandler.InvalidableObjectType;
+
+import static org.keycloak.models.map.common.AbstractMapProviderFactory.MapProviderObjectType.CLIENT_AFTER_REMOVE;
+import static org.keycloak.models.map.common.AbstractMapProviderFactory.MapProviderObjectType.REALM_BEFORE_REMOVE;
+import static org.keycloak.models.map.common.AbstractMapProviderFactory.MapProviderObjectType.ROLE_BEFORE_REMOVE;
 
 /**
  *
  * @author hmlnarik
  */
-public class MapClientProviderFactory extends AbstractMapProviderFactory<ClientProvider, MapClientEntity, ClientModel> implements ClientProviderFactory, ProviderEventListener {
+public class MapClientProviderFactory extends AbstractMapProviderFactory<ClientProvider, MapClientEntity, ClientModel> implements ClientProviderFactory, InvalidationHandler {
 
     private final ConcurrentHashMap<String, ConcurrentMap<String, Integer>> REGISTERED_NODES_STORE = new ConcurrentHashMap<>();
 
-    private Runnable onClose;
-
     public MapClientProviderFactory() {
-        super(ClientModel.class);
+        super(ClientModel.class, ClientProvider.class);
     }
 
     @Override
-    public void postInit(KeycloakSessionFactory factory) {
-        factory.register(this);
-        onClose = () -> factory.unregister(this);
-    }
-
-    @Override
-    public MapClientProvider create(KeycloakSession session) {
+    public MapClientProvider createNew(KeycloakSession session) {
         return new MapClientProvider(session, getStorage(session), REGISTERED_NODES_STORE);
     }
 
-    @Override
-    public void close() {
-        super.close();
-        onClose.run();
-    }
-
     @Override
     public String getHelpText() {
         return "Client provider";
     }
 
     @Override
-    public void onEvent(ProviderEvent event) {
-        if (event instanceof RoleContainerModel.RoleRemovedEvent) {
-            RoleRemovedEvent e = (RoleContainerModel.RoleRemovedEvent) event;
-            RoleModel role = e.getRole();
-            RoleContainerModel container = role.getContainer();
-            RealmModel realm;
-            if (container instanceof RealmModel) {
-                realm = (RealmModel) container;
-            } else if (container instanceof ClientModel) {
-                realm = ((ClientModel) container).getRealm();
-            } else {
-                return;
-            }
-            ((MapClientProvider) e.getKeycloakSession().getProvider(ClientProvider.class)).preRemove(realm, role);
+    public void invalidate(KeycloakSession session, InvalidableObjectType type, Object... params) {
+        if (type == REALM_BEFORE_REMOVE) {
+            ((MapClientProvider) create(session)).preRemove((RealmModel) params[0]);
```

Comment:
Should we check the number of parameters here? Also, is it possible there will be more realms removed (meaning there will be, for example, 10 realm ids in the params argument) with one invalidate call?

Reply:
> Also, is it possible there will be more realms removed (meaning there will be, for example, 10 realm ids in the params argument) with one invalidate call?

I am not sure. It's not happening at the moment, but I can imagine it might be added in future. If we agree I can adapt it to go through all params assuming it's array of `RealmModel`s.